src/java.base/unix/classes/java/lang/ProcessEnvironment.java

```diff
@@ -62,27 +68,92 @@
+                }
+            } else {
+                tracePrunedEnvVarsValue = 0;
+            }
+            // CRIU immutable env var list
+            String propImmutableEnvVars = System.internalGetProperties().getProperty("org.eclipse.openj9.criu.ImmutableEnvVars");
+            String[] immutableEnvVarArray = (propImmutableEnvVars == null) ? new String[] {} : propImmutableEnvVars.split(",");
+            criuImmutableEnvVarList = new HashSet<String>(Arrays.asList(immutableEnvVarArray));
+            // hardcoded list
+            criuImmutableEnvVarList.add("LANG");
+            criuImmutableEnvVarList.add("LC_ALL");
+            criuImmutableEnvVarList.add("LC_CTYPE");
+            criuEnvironment = new HashMap<>(criuImmutableEnvVarList.size() + 3);
+        } else {
+            tracePrunedEnvVarsValue = 0;
+        }
         // We cache the C environment.  This means that subsequent calls
         // to putenv/setenv from C will not be visible from Java code.
         byte[][] environ = environ();
         theEnvironment = new HashMap<>(environ.length/2 + 3);
         // Read environment variables back to front,
         // so that earlier variables override later ones.
+        Variable tmpKeyVar;
         for (int i = environ.length-1; i > 0; i-=2)
-            theEnvironment.put(Variable.valueOf(environ[i-1]),
-                               Value.valueOf(environ[i]));
+        {
+            tmpKeyVar = Variable.valueOf(environ[i-1]);
+            if (criuCapable) {
+                for (String envVar : criuImmutableEnvVarList) {
+                    if (envVar.equals(tmpKeyVar.toString())) {
+                        criuEnvironment.put(tmpKeyVar, Value.valueOf(environ[i]));
+                        break;
+                    }
+                }
+            }
+            theEnvironment.put(tmpKeyVar, Value.valueOf(environ[i]));
+        }
 
-        theUnmodifiableEnvironment
-            = Collections.unmodifiableMap
-            (new StringEnvironment(theEnvironment));
+        if (criuCapable) {
+            theUnmodifiableEnvironment = Collections.unmodifiableMap(new StringEnvironment(criuEnvironment));
+            theOriginalUnmodifiableEnvironment = Collections.unmodifiableMap(new StringEnvironment(theEnvironment));
+        } else {
+            theUnmodifiableEnvironment = Collections.unmodifiableMap(new StringEnvironment(theEnvironment));
+            theOriginalUnmodifiableEnvironment = null;
+        }
     }
 
     /* Only for use by System.getenv(String) */
     static String getenv(String name) {
-        return theUnmodifiableEnvironment.get(name);
+        String currentValue = theUnmodifiableEnvironment.get(name);
+        if (criuCapable && (currentValue == null)) {
+            String origValue = theOriginalUnmodifiableEnvironment.get(name);
+            if (tracePrunedEnvVarsValue == 1) {
+                if (origValue == null) {
+                    System.err.println("No env var set for " + name);
```

Comment:
no need to print or throw if the original value was NULL

Reply:
updated.